### ISSUE TITLE
Aid in aligning colours set in mintty, with colours used in the shell.

### ISSUE THREAD

**Issue by rovf** [mintty/mintty#718]
Basic problem:

(1) mintty allows to configure colour mappings between colour names "Red", "BoldYellow", "CursorColour".

(2) Several tools (such as ls) also allow to configure colours, but in a different way. For instance the colour Red is represented as 31 in the LS_COLOUR variable (http://www.bigsoft.co.uk/blog/index.php/2008/04/11/configuring-ls_colors)

(3) Most shells also have a way to configure the colours used in the Prompt. For instance, Zsh is using colour designators such as %F{magenta}  in their prompt specification.

It would be helpful if we could write a .bashrc or .zshrc file which sets the colours in a way that they work well with the colour set in the respective theme.  Presently, there seems to be no way that I access the mintty theme settings from the shell. Also, it is not always clear, which mintty colour names correspond to which colour name in the shell. To give an example, I can, when configuring th LS_COLOURS variable in my shell, use code 33, which is documented as 'Orange', but there is no Colour named "Orange" in the mintty configuration, and it is therefore not clear, how the effective colour for ls is calculated given a certain mintty theme setting.

While there is probably no way to solve all these problems, I suggest the following enhancements, which could at least help here:

(1) Provide a mintty switch "-E", which, when supplied, would cause mintty to place the current colour settings in to the environment of the spawned process. I suggest environment names corresponding to mintty configuration variables, for instance MINTTY_RED or MINTTY_CURSOR_COLOUR

(2) It would be great if the man-page supplied of mintty also gives examples, how mintty colour names map to common colour configurations such as "ls" or shell-prompts. These are of course only examples, and we can't expect from the mintty developers to collect all this information, but in case some information about this is already known, it would be helpful to have it in the man-page.

**Comment by mintty** [comment]
It's actually quite tricky from the command line because you'd need to fiddle with stty settings and character input timeout.
From `mintheme`, a utility could be derived that does the reverse transformation, if that helps.
I've considered already to extend mintheme to be able to read out configured and current themes.

**Comment by rovf** [comment]
On Thu, Nov 16, 2017, at 22:10, mintty wrote:
> It's actually quite tricky from the command line because you'd need to
> fiddle with stty settings and character input timeout.
> From `mintheme`, a utility could be derived that does the reverse
> transformation, if that helps.

Actually no. Actually, the shell (zsh in this case) being invoked from
within mintty doesn't even know whether the settings where taken from a
theme file (the user could also have set it via command line options),
nor whether the theme file, if any, is in one of the standard locations.
Actually, mintty knows when setting up the terminal, what setting it
uses.

That's why I had suggested the option which generates environment
variables: I think it's easy to understand, it is independent from the
usage of a theme file, and I guess it's not difficult to implement. In
addition, the concept could be extended to make other mintty settings
available to the application. While I don't have any concrete use for
accessing other mintty parameters from the command line, I could for
instance imagine that some people might find it useful to access the -h
option value.

BTW, if you don't like the idea of creating environment entries at the
start of the process, an alternative would be an extension of the
current -R switch. For instance, a 

    mintty -R c

could list the colour settings.


Ronald

**Comment by avih** [comment]
Can you give a very concrete example of what you try to do?

mintty here is the same as other terminals - it let's you define the 16 basic color values (and few more), and the rest is up to each program which runs inside the terminal.

Most program which let you choose colors allow you to choose one of the basic colors or possibly from the fixed 256 colors palette or true colors.

It's two different levels:
1. You as the operator of the terminal can choose the basic palette - the mintty theme.
2. Applications can use (or let you choose) _from that palette_.

**Comment by rovf** [comment]
Example: I have a mintty set up with the configuration

```
BackgroundColour=29,40,37
ForegroundColour=57,113,237
Red=40,42,46
BoldRed=249,106,56
```

I also have a .zshrc (I'm using Zsh, but the basic idea explained here works for Bash too) which has settings like this:

```
PROMPT=$_PBACK'-%F{red}%?%f-%F{magenta}%L%f- %F{green}%U%~%u%f $(git_super_status)'$_PBACK_END'%F{yellow}>%f '
# Bold:1 Dim:2 Underline:4 Reverse:7 Foreground:30-37 Background:40-47
# 30 Black, 31 Red, 32 Green, 33 Orange, 34  Blue, 35  Purple, 36 Cyan, 37  Grey
export LS_COLORS="di=1;4;35:*rc=7:*.rb=32:*.irbrc=32:*.bak=2:*~=2:*.log=32;45:*.txt=34;47:*.sh=36:*.zsh=36"
```

To make the text readable, we need to choose suitable colours for mintty, PROMPT and LS_COLORS (in my example). 

You can see that there is no clear match between the names for the palette. While mintty **and** the PROMPT-expansion **and** the LS_COLORS scheme for example provide the keyword "red", and it is reasonable to assume that they all refer to the same pallette element, when they say *red*, this is not so obvious for other elements. For example, the LS_COLORS specification - which also has pallette of 16 elements - refers to a color name "orange", there is no such color name in the mintty configuration.  On the other hand, mintty provides a "Magenta", which can be found in PROMPT, but not in LS_COLORS. In particular, my request listed in (2) refers to this problem.

Another problem occurs when I write an application which itself creates coloured text, using RGB values. It would help if I know which RGB values were set in the mintty so I can calculate values of good contrast to them. This is the problem my request labelled (1) in my original posting refers to. For instance, if my application knows the current Background Colour, it can calculate a set of suitable contrasting Foreground Colours.

**Comment by avih** [comment]
So the problem is that you don't know how the names map from the application (e.g. orange for LS_COLORS) to the mintty theme?

**Comment by rovf** [comment]
This is one half of the problem. The other half is that I would like to know (from an application running inside a mintty), how the mintty colours are configured. The first is more a documentation problem, while the second one would require a code change in mintty. I now realize that I should have made this two separate issues.

**Comment by avih** [comment]
The first is a documentation problem of the program which lets you choose color names. The 8 colors are standard between all terminals. They are black (0), red (1), green (2), yellow (3), blue (4), magenta/purple (5), cyan (6) and white (7). As you noted at your comment, for foreground colors it's code (SGR) 30-37 of those colors, and for background it's 40-47.

If an application uses colors names other then these 8 codes, then it's a problem with the application, not of mintty. However, if it uses colors from the 256 colors palette or true colours, then there are more names, but here there's no more name conflict because the basic standard is names for only the basic 8 colours.

As for the other problem of reading or setting the theme which mintty uses from an application which runs in the terminal, it's possible too, but not easily and it's generally discouraged. The way to achieve this is via proprietary escape codes which xterm has invented and other terminals have adopted. They're called OSC codes (Operating System Commands - don't ask why), and which mintty supports.

Bottom line, mintty is perfectly normal here and let's you define the standard theme like most other terminals do too. Beyond that, well.. you probably don't wanna get there because modifying the terminal settings from an application can be considered a security issue, and can leave your terminal in an unusable state.

Mintty is fine here.

**Comment by rovf** [comment]
Thanks for the elaborate and understandable explanation!

The only point I didn't quite understand is, why it is difficiult to **read** out the theme, because, if you look at my initial proposal, it could be simply done by setting appropriate environment variables before launching the application; other ways of passing this information are conceivable too.

OTOH, I could understand if there are not many people who would actually use this feature, so in this case, it is of course not worthwile implementing it.

**Comment by avih** [comment]
It's difficult because it's not widely supported and non standard and generally doesn't have practical real world use cases. It's possible in the most standard way there is, which is not very standard. That's it. Your suggestion would add other ways to do the same generally non-practical thing, and there's no reason to do that. If it's really important for you, make the effort and use the way which mintty supports.

**Comment by mintty** [comment]
If you want predictable colours, do not use the "ANSI colours" 0...15.
Use the 6x6x6 colour cube 16...231 and the grey ramp 232...255 (unlike xterm, mintty does not allow them to be reconfigured), or true colour mode if needed.

**Comment by mintty** [comment]
Requesting the current colour configuration from the terminal is easy, just send the escape sequences.
Retrieving the responses and scanning the colour values is fairly easy in an application, unless you want to accomodate for the case of terminals that do not support this and have to implement timeout handling.
Retrieving the responses in a shell script is kind of tricky.

**Comment by rovf** [comment]
Sending escape sequences to receive the colour configuration is possible? This would indeed the most elegant way. How can this be done? I meanwhile found https://en.wikipedia.org/wiki/ANSI_escape_code , but this describes only how to set the colours, not how to retrieve them. 

I guess the response is then sent to /dev/tty?

**Comment by mintty** [comment]
Individual colour configuration (not theme configuration) can be retrieved with the sequences OSC 4 (for 256 colour mode), OSC 5 (for BoldColour), OSC 10 etc (for foreground etc), as described in http://invisible-island.net/xterm/ctlseqs/ctlseqs.html or https://www.xfree86.org/4.8.0/ctlseqs.html .

**Comment by mintty** [comment]
> It is possible to retrieve current colour definitions from mintty and transform them into a desired format.

The new version of the `mintheme` tool will retrieve the current settings and output them in config/theme file format: `mintheme --query` or `mintheme -q`.
Unfortunately, a typo in the 2.8.5 version makes it unusable in the current cygwin package, please download the tool from the [utils repository](https://github.com/mintty/utils).

**Comment by rovf** [comment]
Great! The tool is really helpful.

I am just wondering about the format in which the information is displayed, for instance

    -0-1- ~ > etc/mintheme -q
    ForegroundColour=rgb:c9c9/8383/4444
    BackgroundColour=rgb:2323/2c2c/3131
    CursorColour=rgb:a0a0/3b3b/1e1e
    ...

Why are all colours displayed twice? I would have expected something like

    ForegroundColour=rgb:c9c9c9

or

    ForegroundColour=201,201,201

but this is, of course, a minor issue.

Ronald

**Comment by mintty** [comment]
The tool displays the colour specs in the format as delivered by mintty, when requested via OSC controls.
It could be changed, but it's the same format used by xterm.
Also, the mintheme tool could strip the 4-digit hex format to the 2-digit hex format for display, or even convert to decimal if preferred. Should be a small improvement.

**Comment by mintty** [comment]
Try the attached update (strip the .txt suffix needed for github upload)
[mintheme](https://github.com/mintty/mintty/files/1915812/mintheme.txt)

**Comment by rovf** [comment]
On Mon, Apr 16, 2018, at 16:47, mintty wrote:
> Try the attached update (strip the .txt suffix needed for github upload)
> [mintheme](https://github.com/mintty/mintty/files/1915812/mintheme.txt)

The decimal output looks great now; certainly a useful feature:

ForegroundColour=37,146,79
BackgroundColour=154,165,244
CursorColour=54,199,166

Thanks a lot!

Ronald

**Comment by mintty** [comment]
Released 2.9.0.
